rama-http/src/service/web/router.rs

```diff
@@ -100,7 +49,125 @@ where
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let boxed_service = Arc::new(BoxService::new(service.into_endpoint_service()));
+        match self.routes.insert(path.to_string(), HashMap::new()) {
+            Ok(_) => {
+                if let Some(entry) = self.routes.at_mut(path).ok() {
+                    entry.value.insert(method, boxed_service);
+                }
+            },
+            Err(_err) => {
+                if let Some(existing) = self.routes.at_mut(path).ok() {
+                    existing.value.insert(method, boxed_service);
+                }
+            }
+        };
+        self
+    }
+
+    pub fn get<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::GET, path, service)
+    }
+
+    pub fn post<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::POST, path, service)
+    }
+
+    pub fn put<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::PUT, path, service)
+    }
+
+    pub fn delete<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::DELETE, path, service)
+    }
+
+    pub fn patch<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::PATCH, path, service)
+    }
+
+    pub fn head<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::HEAD, path, service)
+    }
+
+    pub fn options<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::OPTIONS, path, service)
+    }
+
+    pub fn trace<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.route(Method::TRACE, path, service)
+    }
+
+    /// use the given service in case no match could be found.
+    pub fn not_found<I, T>(mut self, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.not_found = Arc::new(service.into_endpoint_service().boxed());
+        self
+    }
+}
+
+impl<State> Service<State, Request> for Router<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    type Response = Response;
+    type Error = Infallible;
+
+    async fn serve(
+        &self,
+        mut ctx: Context<State>,
+        req: Request<>,
+    ) -> Result<Self::Response, Self::Error> {
+        let uri_string = req.uri().to_string();
+        match &self.routes.at(uri_string.as_str()) {
+            Ok(matched) => {
+                let params: HashMap<String, String> = matched.params.clone().iter().map(|(k, v)| (k.to_string(), v.to_string())).collect();
```

Comment:
This will need to be collected as https://docs.rs/rama/latest/rama/http/matcher/struct.UriParams.html to be compatible with the existing matcher object. Feel free to expand the logic of UriParams if you miss something.

E.g. might make code cleaner here if you implement `FromIterator` for `UriParams`, but that's more of an optional thing.

Reply:
Also in general there should be no reason to double allocate.

1. first allocation is .clone()
2. than you allocate again by turning the pairs into owned pairs

Probably that first clone isn't needed